qkeras/quantizers.py

```diff
@@ -439,9 +440,537 @@ def trainable_variables(self):
+      quantization_scale = self.quantization_scale
+
+    scaled_xq = self._scale_clip_and_round(x, quantization_scale)
+    xq = scaled_xq * quantization_scale
+
+    res = x + self.qnoise_factor * (xq - x)
+
+    return res
+  
+  def _scale_clip_and_round(self, x, quantization_scale):
+    """Scale, clip, and round x to an integer value in a limited range
+    Note that the internal shift is needed for 1-bit quantization to ensure 
+    that a sign function is used."""
+
+    # special shifting needed to compute a sign function.
+    shift = self.use_sign_function * 0.5
+
+    clip_min, clip_max = self.clip_bounds
+
+    scaled_x = x / quantization_scale
+    clipped_scaled_x = K.clip(scaled_x, clip_min, clip_max)
+    # Round through to nearest integer, using straight-through estimator 
+    # for gradient computations. 
+    scaled_xq = _round_through(
+      clipped_scaled_x - shift,
+      use_stochastic_rounding=self.use_stochastic_rounding,
+      precision=1.0, # using 1.0 precision so that we round to a nearby integer
+    )
+
+    return scaled_xq + shift
+    
+  def _get_quantization_scale(self, x):
+    """Get quantization_scale, either from self or from input x"""
+
+    # Get the minimum floating point scale that does not clip the max of x
+    # This is the quantization scale for alpha="auto"
+    quantization_scale = self._get_quantization_scale_from_max_data(x)
+
+    if self.alpha == "auto_po2":
+      quantization_scale = self._po2_autoscale(x, quantization_scale)
+
+    # update quantization_scale variable
+    # stop_gradient on quantization_scale to ignore dependence on x
+    self.quantization_scale = tf.stop_gradient(quantization_scale)
+
+    # very important that return value is a tf.Variable with shape None
+    return self.quantization_scale 
+
+  def _get_quantization_scale_from_max_data(self, x):
+    """Get the minimum floating point scale that does not clip the max 
+    of x"""
+
+    axis = _get_scaling_axis(self.scale_axis, tf.rank(x))
+
+    clip_min, clip_max = self.clip_bounds
+    clip_range = clip_max - clip_min
+    
+    if self.keep_negative:
+      data_max = K.max(tf.math.abs(x), axis=axis, keepdims=True)
+      quantization_scale = (data_max * 2) / clip_range
+    else:
+      data_max = K.max(x, axis=axis, keepdims=True)
+      quantization_scale = data_max / clip_range
+
+    return tf.math.maximum(quantization_scale, K.epsilon())
+
+  def _po2_autoscale(self, x, quantization_scale):
+    """Get an approximation of the "best" po2 scale using least squares"""
+
+    # set alpha scale to a near power of two
+    quantization_scale = K.pow(2.0, 
+                         tf.math.round(K.log(quantization_scale + K.epsilon()) / 
+                                       K.log(2.0)))
+
+    def loop_body(_, quantization_scale):
+      """Loop body for least squares autoscaling"""
+
+      scaled_xq = self._scale_clip_and_round(x, quantization_scale)
+      new_quantization_scale = _get_scale(
```

Comment:
Should we rename this function to something more obvious?

Reply:
I went through a few iterations of name changes here, but this one seemed the most specific to me. I'm very open to suggestions

Reply:
Renamed to _get_least_squares_scale